libs/plugin.js

```diff
@@ -31,25 +33,84 @@ class DrupalJsonApi {
     if (this.isCached(endpoint)) {
       return Promise.resolve(this.getCached(endpoint))
     }
-    this.pending.add(endpoint)
-    let res = false
-    try {
-      res = await this.api.get(endpoint)
-    } catch (err) {
-      if (this.isEntity(err.response)) {
-        res = err.response
-      } else if (err.response.status === 403) {
-        res = apiError(403, 'Not Authorized')
-      } else if (err.response.status === 404) {
-        res = apiError(404, 'Not Found')
-      } else {
-        res = apiError()
+    this.pending.set(endpoint, new Promise(async (resolve) => {
+      let res = false
+      try {
+        res = await this.api.get(endpoint)
+      } catch (err) {
+        if (err.response) {
+          if (this.isEntity(err.response)) {
+            res = err.response
+          } else if (err.response.status === 403) {
+            res = apiError(403, 'Not Authorized')
+          } else if (err.response.status === 404) {
+            res = apiError(404, 'Not Found')
+          } else {
+            res = apiError()
+          }
+        } else {
+          console.error('bad request: ', endpoint, err)
+          res = apiError()
+        }
       }
+      const d = this.isEntity(res) ? (new DrupalJsonApiEntity(this, res.data)) : res.data
+      this.setCache(endpoint, d)
+      resolve(d)
+    }))
+    const entity = await this.pending.get(endpoint)
+    this.pending.delete(endpoint)
+    return entity
+  }
+
+  /**
+   * Retrieve a given endpoint (from cache or filesystem)
+   * @param {string} endpoint
+   * @return {Promise}
+   */
+  async fromFileSystem (endpoint) {
+    if (this.isCached(endpoint)) {
+      return Promise.resolve(this.getCached(endpoint))
     }
-    const d = this.isEntity(res) ? (new DrupalJsonApiEntity(this, res.data)) : res.data
-    this.setCache(endpoint, d)
+    this.fs = this.fs ? this.fs : __non_webpack_require__('fs')
+    const fsEndpoint = './dist/api' + endpoint
+    this.pending.set(endpoint, new Promise((resolve) => {
+      this.fs.readFile(fsEndpoint, 'utf8', (err, data) => {
+        const res = err ? apiError() : {data: JSON.parse(data)}
+        const d = this.isEntity(res) ? (new DrupalJsonApiEntity(this, res.data)) : res
+        this.setCache(endpoint, d)
+        resolve(d)
+      })
+    }))
+    const entity = await this.pending.get(endpoint)
     this.pending.delete(endpoint)
-    return d
+    return entity
+  }
+
+  /**
+   * Retrieve a given endpoint (from cache or filesystem)
+   * @param {string} endpoint
+   * @return {Promise}
+   */
+  fromFileSystemBySlug (lookup) {
+    this.fs = this.fs ? this.fs : __non_webpack_require__('fs')
+    const endpoint = `./dist/api/_slugs${this.trimSlug(lookup.slug)}.json`
+    return new Promise((resolve, reject) => {
+      this.fs.readFile(endpoint, 'utf8', async (err, data) => {
+        if (err) {
+          reject(err)
```

Comment:
This seems like a possible failure point in the future, we're rejecting here but then not really catching it or doing anything else to gracefully remedy the situation. Not entirely sure what a better alternative is though. No change necessary on this one.